src/esmockModule.js

```diff
@@ -13,7 +13,13 @@ import {
 const isObj = o => typeof o === 'object' && o
 const isDefaultIn = o => isObj(o) && 'default' in o
 const isDirPathRe = /^\.?\.?([a-zA-Z]:)?(\/|\\)/
+const isMetaResolve = typeof import.meta.resolve === 'function'
 const nextId = ((id = 0) => () => ++id)()
+const rootDirSlashRe = /^\//
+const protocolFileRe = /^file:/
+
+const asFileURL = p => protocolFileRe.test(p) ? p
+  : `file:///${p.replace(rootDirSlashRe, '')}`
```

Comment:
it's need to use fileURLToPath from 'url' to convert regular path to url. Just because encoding, spaces, etc.

Reply:
The tests include paths with spaces and tests are passing https://github.com/iambumblehead/esmock/blob/master/tests/tests-node/esmock.node.test.js#L386-L394

If you can show an example scenario or test where this function fails, I will change it

Reply:
https://nodejs.org/dist/latest-v19.x/docs/api/esm.html#import-specifiers
"Bare specifier resolutions are handled by the Node.js module resolution algorithm. All other specifier resolutions are always only **resolved with the standard relative URL** resolution semantics"

Fastests example (may be not related to current case but related to paths / URL processing):
<img width="1172" alt="image" src="https://user-images.githubusercontent.com/20106607/196745547-027e413b-75f0-41a0-a1e5-a96f47bf6f6d.png">
<img width="703" alt="image" src="https://user-images.githubusercontent.com/20106607/196745650-8fcfd69d-4366-4023-894a-898985df563b.png">

I found the similar issues in my 'ts loader for Yarn PnP' project - until I started treating all incoming specifiers as URLs (with possible ? or # appended), relative or absolute.

Reply:
Would you rewrite and attach the function definition you think is best? It is OK if it doesn't work perfectly, I can run tests here and make adjustments and push things up after tests are passing.

I want to learn which way is safest from your experience

Reply:
esmock would not handle urls with query strings as in the screenshot. If that situation needs to be handled, changes could be added in a separate MR.

Let's wait and support that situation later.

Reply:
First, how to test it. Due to you detect parent path via new Error() stack, we have to load test file a bit special:
<img width="554" alt="image" src="https://user-images.githubusercontent.com/20106607/196757084-8fd8a87d-82be-4d59-88cd-e6298181026c.png">

Then we can call `node --experimental-import-meta-resolve --loader=esmock ./wrapper.js`, without any existing code modifications. Something failed:
<img width="108" alt="image" src="https://user-images.githubusercontent.com/20106607/196757468-9183a9d0-960f-4d4b-acdd-a9faa9ee9302.png">
<img width="628" alt="image" src="https://user-images.githubusercontent.com/20106607/196757646-e2d639b5-f69b-4b63-bfc8-1db34e6995d8.png">

It fails also without meta resolve via `node --loader=esmock ./wrapper.js`.

And this issue (bad re in test, 95%) is not related to 'asFileUrl' issue - because in *nix systems you will always get absolute path (or file url) in stack trace, see my simple trace from asFileUrl:
<img width="113" alt="image" src="https://user-images.githubusercontent.com/20106607/196759617-6abeb02d-a17a-40f5-ae62-fe5b58c641ca.png">

So it's a hacky but safe code at the moment when you use stacktrace to get parent path. My suggestion that you asked:
<img width="623" alt="image" src="https://user-images.githubusercontent.com/20106607/196760766-76e0ba56-efa7-48de-bfce-3051074803c8.png">
`npm run test:metaresolve`
<img width="106" alt="image" src="https://user-images.githubusercontent.com/20106607/196761392-8e02103e-60fc-41e8-bd89-5695c09127e5.png">

Reply:
ok thanks for the great explanation. will add this now.

Reply:
@koshic the change is applied and tests are passing, closing this as resolved